galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/OidcProvider.java

```diff
@@ -46,40 +55,124 @@ public class OidcProvider {
+        session.setAttribute("state", state);
+
+        String authUrl = authorizationEndpoint + queryParams;
+        return sendGetRequest(URI.create(authUrl));
     }
 
     /**
      * Sends a POST request with a given request body to an OpenID Connect /token endpoint, returning the received response.
      */
-    public HttpResponse<String> sendTokenPost(JsonObject requestBody) throws IOException, InterruptedException {
+    public HttpResponse<String> sendTokenPost(String clientId, String clientSecret, String refreshToken) throws IOException, InterruptedException {
 
         StringBuilder sbRequestBody = new StringBuilder();
-        String clientId = requestBody.get("client_id").getAsString();
-        String refreshToken = requestBody.get("refresh_token").getAsString();
-        String secret = requestBody.get("secret").getAsString();
-        String decodedSecret = new String(Base64.getDecoder().decode(secret));
 
         // Convert the request's JSON object to the application/x-www-form-urlencoded content type
         // as required by the /token endpoint.
         sbRequestBody.append("grant_type=refresh_token");
         sbRequestBody.append("&client_id=" + clientId);
-        sbRequestBody.append("&client_secret=" + decodedSecret);
+        sbRequestBody.append("&client_secret=" + clientSecret);
         sbRequestBody.append("&refresh_token=" + refreshToken);
 
         // Create a POST request to the /token endpoint
-        HttpRequest postRequest = HttpRequest.newBuilder()
-            .POST(BodyPublishers.ofString(sbRequestBody.toString()))
-            .header("Content-type", "application/x-www-form-urlencoded")
-            .uri(URI.create(issuerUrl + "/token"))
-            .build();
+        return sendPostRequest(sbRequestBody.toString(), "application/x-www-form-urlencoded", URI.create(tokenEndpoint));
+    }
+
+    /**
+     * Sends a POST request with a given request body to an OpenID Connect /token endpoint, returning the received response.
+     */
+    public HttpResponse<String> sendTokenPost(String clientId, String clientSecret, String authCode, String redirectUri) throws IOException, InterruptedException {
+
+        StringBuilder sbRequestBody = new StringBuilder();
 
-        HttpResponse<String> response = httpClient.send(postRequest, BodyHandlers.ofString());
-        return response;
+        // Convert the request's JSON object to the application/x-www-form-urlencoded content type
+        // as required by the /token endpoint.
+        sbRequestBody.append("grant_type=authorization_code");
+        sbRequestBody.append("&code=" + authCode);
+        sbRequestBody.append("&client_id=" + clientId);
+        sbRequestBody.append("&client_secret=" + clientSecret);
+        sbRequestBody.append("&redirect_uri=" + redirectUri);
+
+        // Create a POST request to the /token endpoint
+        return sendPostRequest(sbRequestBody.toString(), "application/x-www-form-urlencoded", URI.create(tokenEndpoint));
```

Comment:
If this fails, we won't see any logging ? Might be good to add more logging here and there perhaps.